File: go/aead/aead.go
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

// Package aead provides implementations of the AEAD primitive.
//
// AEAD encryption assures the confidentiality and authenticity of the data. This primitive is CPA secure.
package aead

import (
	"fmt"

	"github.com/google/tink/go/core/registry"
	"github.com/google/tink/go/internal/internalapi"
	"github.com/google/tink/go/internal/internalregistry"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
)

// Config is an interface used by key managers and primitive factories to
// represent the relevant Config functionality.
//
// Note that the interface is not public and meant for package-internal use
// only, despite it being exported.
//
// TODO: b/286235179 -- link the implementation.
type Config interface {
	RegisterKeyManager(km registry.KeyManager, t internalapi.Token) error
	PrimitiveFromKeyData(keyData *tinkpb.KeyData, _ internalapi.Token) (any, error)
}

func init() {
	if err := registry.RegisterKeyManager(new(aesCTRHMACAEADKeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}

	if err := registry.RegisterKeyManager(new(aesGCMKeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}
	if err := internalregistry.AllowKeyDerivation(aesGCMTypeURL); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}

	if err := registry.RegisterKeyManager(new(chaCha20Poly1305KeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}

	if err := registry.RegisterKeyManager(new(xChaCha20Poly1305KeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}
	if err := internalregistry.AllowKeyDerivation(xChaCha20Poly1305TypeURL); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}

	if err := registry.RegisterKeyManager(new(kmsEnvelopeAEADKeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}

	if err := registry.RegisterKeyManager(new(aesGCMSIVKeyManager)); err != nil {
		panic(fmt.Sprintf("aead.init() failed: %v", err))
	}
}

File: python/aead/aead.py
# Copyright 2019 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This module defines the interface for AEAD."""

from __future__ import absolute_import
from __future__ import division
# Placeholder for import for type annotations
from __future__ import print_function

import abc

# Special imports
import six


@six.add_metaclass(abc.ABCMeta)
class Aead(object):
  """The interface for authenticated encryption with associated data.

  Implementations of this interface are secure against adaptive
  chosen ciphertext attacks.  Encryption with associated data ensures
  authenticity and integrity of that data, but not its secrecy.
  (see RFC 5116, https://tools.ietf.org/html/rfc5116)
  """

  @abc.abstractmethod
  def encrypt(self, plaintext: bytes, associated_data: bytes) -> bytes:
    """Encrypts plaintext with associated_data.

    The ciphertext allows for checking authenticity and integrity of the
    associated data, but does not guarantee its secrecy.

    Args:
      plaintext: bytes. The data to be encrypted.
      associated_data: bytes. The associated data, that will be authenticated.
    Returns:
      the resulting ciphertext as bytes.
    Raises:
      tink.TinkError if the encryption fails.
    """
    raise NotImplementedError()

  @abc.abstractmethod
  def decrypt(self, ciphertext: bytes, associated_data: bytes) -> bytes:
    """Decrypts ciphertext with associated_data.

    The decryption verifies the authenticity and integrity of the associated
    data, but there are no guarantees with respect to secrecy of that data.

    Args:
      ciphertext: bytes. The data to be decrypted.
      associated_data: bytes. The associated data.
    Returns:
      the resulting plaintext as bytes.
    Raises:
      tink.TinkError if the decryption fails.
    """
    raise NotImplementedError()
